File: Remote.java
import java.nio.charset.StandardCharsets;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

public class Remote {
    private static final int PORT_NUMBER = 1099;

    public static void main(String[] args) {
        String host = System.getenv("RPC_HOST");
        String port = System.getenv("RPC_PORT");
        int portNumber = PORT_NUMBER;
        if (port != null) {
            portNumber = Integer.parseInt(port);
        }
        try {
            Registry registry = LocateRegistry.getRegistry(host, portNumber);
            RPCStubs stub = (RPCStubs) registry.lookup("RPCStubs");

            System.out.println("File size (in bytes): " + stub.size("test_file.txt"));
            SimpleDateFormat sdf = new SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss z");
            sdf.setTimeZone(TimeZone.getTimeZone("GMT"));
            System.out.println("Last modified time: " + sdf.format(
                new Date(stub.lastModifiedTime("test_file.txt"))));
            int fd = stub.open("test_file.txt", "rw");
            byte[] buf = new byte[200];
            byte[] resp = stub.read(fd, 5);
            System.arraycopy(resp, 0, buf, 0, 5);
            System.out.println(resp.length + " " + new String(buf, StandardCharsets.UTF_8));
            stub.seek(fd, 0);
            byte[] resp2 = stub.read(fd, 6);
            System.arraycopy(resp2, 0, buf, 5, 6);
            System.out.println(resp2.length + " " + new String(buf, StandardCharsets.UTF_8));
            stub.write(fd, "word".getBytes(StandardCharsets.UTF_8), 0, 4);
            stub.close(fd);
        } catch (Exception e) {
            System.err.println("TestRemote exception: " + e.toString());
            e.printStackTrace();
        }
    }
}

File: remote.py
import os
import rpyc
import time

PORT_NUMBER = 5000

host = os.getenv("RPC_HOST")
port = os.getenv("RPC_PORT")
if host is None:
    host = "127.0.0.1"
if port is None:
    port_number = PORT_NUMBER
else:
    port_number = int(port)

conn = rpyc.connect(host, port_number)

statbuf = conn.root.stat("test_file.txt")
statbuf = os.stat_result(statbuf)
print("File size (in bytes): " + str(statbuf.st_size))
print("Last modified time: " + time.strftime("%a, %d %b %Y %X GMT",
                                             time.gmtime(int(statbuf.st_mtime))))
fd: int = conn.root.open("test_file.txt", os.O_RDWR)
buf: bytes = conn.root.read(fd, 5)
print(str(len(buf)) + " " + buf.decode())
conn.root.lseek(fd, 0, os.SEEK_SET)
buf2: bytes = conn.root.read(fd, 6)
print(str(len(buf2)) + " " + buf.decode() + buf2.decode())
status = conn.root.write(fd, "word".encode())
conn.root.close(fd)
